### PR TITLE
[03] Implement the queue-handler

### PR DESCRIPTION
**Pull request by rexwangcc**
This PR implements the queue handler to retrieve workflows from Cromwell.

### CODE REVIEW

File: falcon/queue_handler.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    def __str__(self):
+        return str(self.id)
+
+    def __repr__(self):
+        return str(self.id)
+
+    def __eq__(self, other):
+        if isinstance(other, Workflow):
+            return self.id == other.id and self.bundle_uuid == other.bundle_uuid
```
Comment (samanehsan) [inline falcon/queue_handler.py]: If `self.id == other.id` I don't think it's necessary to check that `self.bundle_uuid == other.bundle_uuid`

File: falcon/queue_handler.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        ).json()['results']
+
+        logger.info('Queue | Retrieved {0} workflows from Cromwell. | {1}'.format(len(workflows), datetime.now()))
+
+        # store the latest submission timestamp to save computation time for later updates
+        self.last_submission = workflows[-1].get('submission')
+
+        # placeholder for the de-duplication logic
+        for workflow in workflows:
+            workflow = Workflow(workflow.get('workflow_id'), workflow.get('bundle-uuid'))
```
Comment (samanehsan) [inline falcon/queue_handler.py]: Should this be `workflow.get('bundle_uuid'))`?

File: falcon/queue_handler.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+        while True:
+            if self.last_submission:
+                # make sure it only queries the workflows submitted after the last retrieve, if applicable
+                self.cromwell_query_dict['submission'] = self.last_submission  # this is not a atomic manipulation!
+            self.retrieve_queue(self.cromwell_query_dict)
+            self.sleep_for(self.settings.get('queue_update_interval'))
+
+    @staticmethod
+    def shallow_deduplicate(ls):
```
Comment (samanehsan) [inline falcon/queue_handler.py]: What is `ls` short for?

Comment (rexwangcc) [reply falcon/queue_handler.py]: The `ls` here refers to "list", which is kinda outdated in the code, will change it to item, nice point!

File: falcon/queue_handler.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    def retrieve_queue(self, query_dict):
+        """Retrieve the latest set of "Ob Hold" workflows from Cromwell and put them in the in-memory queue.
+
+        Args:
+            query_dict (dict): A dictionary that contains valid query parameters which can be accepted by the Cromwell
+            /query  endpoint.
+        """
+        # workflows should be a list that ordered by submission time, oldest first
+        workflows = cromwell_tools.query_workflows(
```
Comment (samanehsan) [inline falcon/queue_handler.py]: In Cromwell 34, the order of the workflows returned will change from newest --> oldest!

File: falcon/queue_handler.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        time.sleep(sleep_time)
+
+    def retrieve_queue(self, query_dict):
+        """Retrieve the latest set of "Ob Hold" workflows from Cromwell and put them in the in-memory queue.
+
+        Args:
+            query_dict (dict): A dictionary that contains valid query parameters which can be accepted by the Cromwell
+            /query  endpoint.
+        """
+        # workflows should be a list that ordered by submission time, oldest first
```
Comment (samanehsan) [inline falcon/queue_handler.py]: In Cromwell 34, the order of the workflows returned will change from newest --> oldest!

File: falcon/queue_handler.py (hunk 2 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: This field refers to the JSON query result from Cromwell, so it should actually be `additionalQueryResultFields.labels.bundle-uuid` according to the description of the `/query` endpoint of Cromwell, will change it.

File: falcon/queue_handler.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            query_dict=query_dict,
+            cromwell_user=self.settings.get('cromwell_user'),
+            cromwell_password=self.settings.get('cromwell_password'),
+            caas_key=self.settings.get('caas_key')
+        ).json()['results']
+
+        logger.info('Queue | Retrieved {0} workflows from Cromwell. | {1}'.format(len(workflows), datetime.now()))
+
+        # store the latest submission timestamp to save computation time for later updates
+        self.last_submission = workflows[-1].get('submission')
```
Comment (samanehsan) [inline falcon/queue_handler.py]: This will throw an error if workflows is an empty list

File: falcon/queue_handler.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        self.process.start()
+
+    def join(self):
+        self.process.join()
+
+    def sleep_for(self, sleep_time):
+        time.sleep(sleep_time)
+
+    def retrieve_queue(self, query_dict):
+        """Retrieve the latest set of "Ob Hold" workflows from Cromwell and put them in the in-memory queue.
```
Comment (samanehsan) [inline falcon/queue_handler.py]: nit: typo in "Ob Hold" --> should be "On Hold"

File: falcon/queue_handler.py (hunk 1 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: Yes you are right. I added `self.bundle_uuid == other.bundle_uuid` here when I thought I will also implement the de-duplication logic, but since it doesn't make any sense currently, I'm going to remove it for now.

File: falcon/queue_handler.py (hunk 2 above)
Comment (samanehsan) [reply falcon/queue_handler.py]: Oh, I see `workflow` here refers to the workflow JSON returned from Cromwell!

File: falcon/queue_handler.py (hunk 5 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: Oh! That's an important fact to know! Thanks!

Comment (rexwangcc) [reply falcon/queue_handler.py]: Speaking of that, do you know if there is a way to control the order of the workflows in the query params? If not, it will be awkward for falcon, and we have to add some comparasion between `workflows[-1].get('submission')` and `workflows[0].get('submission')` before chaning the state.

File: falcon/queue_handler.py (hunk 6 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: Nice catch! Will fix

File: falcon/queue_handler.py (hunk 7 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: Oops typo 😅 Thanks for catching!

File: falcon/queue_handler.py (hunk 2 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: Do you think the variable name here is confusing since the workflows in the mem-queue are also called workflow, should I change it to something like `workflow_from_cromwell`?

File: falcon/queue_handler.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+        # placeholder for the de-duplication logic
+        for workflow in workflows:
+            workflow = Workflow(workflow.get('workflow_id'), workflow.get('bundle-uuid'))
+
+            logger.debug(
+                'Queue | Enqueuing workflow {0} | {1}'.format(workflow, datetime.now()))
+
+            self.workflow_queue.put(self.deep_deduplicate(
+                self.shallow_deduplicate(workflow)
```
[the remark this answers is not in the document]
Comment (rexwangcc) [reply falcon/queue_handler.py]: Nice point, we need to think carefully about the de-dup logic here, since it will very likely become the bottleneck of the falcon in the future. For now, do you mind if I add more comments here to clarify this and leave the 2 de-dup functions as identity functions?

[the remark this answers is not in the document]
Comment (samanehsan) [reply falcon/queue_handler.py]: I think it is clearer to do `self.workflow_queue.put(workflow)` and add a TODO to deduplicate it, instead of calling the placeholder functions.

[the remark this answers is not in the document]
Comment (rexwangcc) [reply falcon/queue_handler.py]: That makes sense, will do!

File: falcon/queue_handler.py (hunk 5 above)
Comment (samanehsan) [reply falcon/queue_handler.py]: I think it just reverses the order, but Chris on the Cromwell team should be able to confirm what the behavior will be

File: falcon/queue_handler.py (hunk 2 above)
Comment (samanehsan) [reply falcon/queue_handler.py]: Maybe call it `workflow_metadata`?

File: falcon/queue_handler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (samanehsan) [reply falcon/queue_handler.py]: Thanks :)

File: falcon/queue_handler.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            self.process = Process(target=self.execution)
+        self.process.start()
+
+    def join(self):
+        self.process.join()
+
+    def sleep_for(self, sleep_time):
+        time.sleep(sleep_time)
+
+    def retrieve_queue(self, query_dict):
```
Comment (samanehsan) [inline falcon/queue_handler.py]: Maybe rename this to retrieve_and_queue_workflows? Since it's both getting workflows from Cromwell and adding them to a queue

File: falcon/queue_handler.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger('falcon.{module_path}'.format(module_path=__name__))
+
+
+class Workflow(object):
+    """A simple data structure for hosting workflow data.
+
+    Besides the features for de-duplication, this class also utilizes a smaller size of chunk in memory,
```
Comment (samanehsan) [inline falcon/queue_handler.py]: I think you can remove the part that compares this class to other data structures

File: falcon/queue_handler.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+        if response.status_code != 200:
+            logger.warning('Queue | Failed to retrieve workflows from Cromwell | {0}'.format(datetime.now()))
+            logger.info('Queue | {0} | {1}'.format(response.text, datetime.now()))
+        else:
+            workflow_metas = response.json()['results']
+
+            # TODO: This count-inconsistency issue has been fixed after Cromwell v34
+            results_count = response.json()['totalResultsCount']
+            workflow_num = results_count if results_count == len(workflow_metas) else len(workflow_metas)
```
Comment (samanehsan) [inline falcon/queue_handler.py]: How about just doing `workflow_num = len(workflow_metas)` here? It looks like you need to check the length of the list anyways to use the results_count since it may be incorrect

File: falcon/queue_handler.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            logger.info('Queue | Retrieved {0} workflows from Cromwell. | {1}'.format(workflow_num, datetime.now()))
+
+            if workflow_num:
+                # store the latest submission timestamp to save computation time for later updates
+                # TODO: from Cromwell v34 (https://github.com/broadinstitute/cromwell/releases/tag/34), Query results will
+                # be returned in reverse chronological order, with the most-recently submitted workflows returned first,
+                # the logic here need to be updated.
+                self.last_submission = workflow_metas[-1].get('submission')
+
+                # placeholder for the de-duplication logic
```
Comment (samanehsan) [inline falcon/queue_handler.py]: Please remove this old comment

File: falcon/queue_handler.py [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+                        "id": "xxx2",
+                        "submission": "2018-01-01T23:49:42.171Z",
+                        "status": "Succeeded",
+                        "end": "2018-07-12T00:31:27.273Z",
+                        "start": "2018-07-11T23:49:48.385Z"
+                    }
+                ]
+            ```
+        """
+        workflow_num = len(workflow_metas)
```
Comment (dshiga) [inline falcon/queue_handler.py]: num_workflows would be a little less open to interpretation

File: falcon/queue_handler.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+            'Queue | Initialing the queue handler with thread => {0} | {1}'.format(get_ident(), datetime.now()))
+
+        while True:
+            self.enqueue(
+                self.retrieve_workflows(self.cromwell_query_dict)
+            )
+            self.sleep_for(self.queue_update_interval)
+
+    @staticmethod
+    def is_workflow_list_in_oldest_first_order(workflow_list):
```
Comment (dshiga) [inline falcon/queue_handler.py]: Would be good to add unit tests for this and have them guarantee that it still works when the list is empty or has just one element in it.

File: falcon/queue_handler.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+
+                self.workflow_queue.put(workflow)  # TODO: Implement and add de-duplication logic here
+        else:
+            logger.info(
+                'Queue | Cannot fetch any workflow from Cromwell, go back to sleep and wait for next attempt. | {0}'.format(
+                    datetime.now()))
+
+    def execution(self):
+        logger.info(
+            'Queue | Initialing the queue handler with thread => {0} | {1}'.format(get_ident(), datetime.now()))
```
Comment (dshiga) [inline falcon/queue_handler.py]: Initialize

File: falcon/queue_handler.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+    def __repr__(self):
+        return str(self.id)
+
+    def __eq__(self, other):
+        if isinstance(other, Workflow):
+            return self.id == other.id
+        return False
+
+
+class Queue_Handler(object):
```
Comment (dshiga) [inline falcon/queue_handler.py]: I think standard style is no underscores for class names -> QueueHandler.

File: falcon/queue_handler.py [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+            caas_key=self.settings.get('caas_key')
+        )
+
+        if response.status_code != 200:
+            logger.warning('Queue | Failed to retrieve workflows from Cromwell | {0} | {1}'.format(response.text, datetime.now()))
+            return []
+        else:
+            return response.json()['results']
+
+    def enqueue(self, workflow_metas):
```
Comment (dshiga) [inline falcon/queue_handler.py]: I would reorganize a bit. Why not have retrieve_workflows return a list of Workflow objects rather than creating them in this function? I would also put the workflow order check / reversal code in retrieve_workflows, or else create separate functions for these things that you call prior to enqueue -- so that enqueue() can be very simple and fast.

I would also pass in  the queue to the enqueue function, rather than creating it inside enqueue -- this should make it easier to test (and also avoid surprising users, since it might be a bit surprising that the enqueue() function actually discards the old queue and creates a new one).

So if you follow the above you'd end up with code more like:
```
workflows = self.retrieve_workflows(self.cromwell_query_dict)
self.enqueue(workflows, self.get_queue())
```

File: falcon/queue_handler.py [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+        self.thread.start()
+
+    def join(self):
+        try:
+            self.thread.join()
+        except (AttributeError, AssertionError):
+            logger.error('The thread of this queue handler is not in a running state.')
+
+    @staticmethod
+    def obtainQueue(max_queue_size=-1):
```
Comment (dshiga) [inline falcon/queue_handler.py]: I'm not sure this really needs a function of its own, since it's really just calling through to `Queue()`. If you do need it for some reason, I would probably call it `create_empty_queue()` to be more explicit about what it's doing.

Comment (rexwangcc) [reply falcon/queue_handler.py]: This is implemented as a factory function which I thought would be useful if in the future we want to change to use other `Queue` implementations besides `queue.Queue()` (such as `asyncio.Queue` or some 3rd party `Queue`s).  `create_empty_queue()` sounds like a good choice to me!

File: falcon/queue_handler.py (hunk 17 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: That makes more sense to me. The only question is how to tell `queue_handler()` keep the current `queue` object when `self.retrieve_workflows()` retrieve 0 workflow from the Cromwell. Adding `if` logic for this might complicate the `execution()`, which I thought was problematic before but feel trivial now. Will make the change.

File: falcon/queue_handler.py [hunk 19]
```diff
@@ -122,49 +149,103 @@ def enqueue(self, workflow_metas):
...
+            workflows (iterable): An iterable(list or iterator) object that contains all `Workflow` instances that need
+            to be put in the in-memory queue.
+        """
+        for workflow in workflows:
+            logger.debug('Queue | Enqueuing workflow {0} | {1}'.format(workflow, datetime.now()))
+            self.workflow_queue.put(workflow)  # TODO: Implement and add de-duplication logic here
 
-                logger.debug(
-                    'Queue | Enqueuing workflow {0} | {1}'.format(workflow, datetime.now()))
+    def rebuild_queue(self, queue):
```
Comment (dshiga) [inline falcon/queue_handler.py]: I think `set_queue` would be a more accurate and standard name for this type of function.

File: falcon/queue_handler.py [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+            head = datetime.strptime(str(workflow_list[0].get('submission')), CROMWELL_DATETIME_FORMAT)
+            tail = datetime.strptime(str(workflow_list[-1].get('submission')), CROMWELL_DATETIME_FORMAT)
+            if head <= tail:
+                return True
+        except ValueError:
+            logger.error(
+                'Queue | An error happened when try to parse the submission timestamps, will assume oldest first for'
+                ' the workflows returned from Cromwell | {0}'.format(datetime.now()))
+            return True
+        return False
```
Comment (samanehsan) [inline falcon/queue_handler.py]: You can consolidate this by doing `return head <= tail` here, and removing the check on lines 287-288

File: falcon/queue_handler.py [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+        logger.info(
+            'Queue | Initializing the queue handler with thread => {0} | {1}'.format(get_ident(), datetime.now()))
+
+        while True:
+            workflow_metas = self.retrieve_workflows(self.cromwell_query_dict)
+            if workflow_metas:  # This could happen when getting either non-200 codes or 0 workflow from Cromwell
+                workflows = self.prepare_workflows(workflow_metas)
+
+                self.rebuild_queue(self.create_empty_queue(-1))  # This must happen before `enqueue()` is called!
+                self.enqueue(workflows)
```
Comment (dshiga) [inline falcon/queue_handler.py]: How about:
```
queue = self.create_empty_queue(-1)
self.enqueue(workflows, queue)
self.set_queue(queue)
```
That way the igniter will continue using the old queue until the new queue is fully populated. Then we switch to the new queue in a single atomic operation. Passing the queue as an argument to self.enqueue might also facilitate testing.

Comment (rexwangcc) [reply falcon/queue_handler.py]: I thought about doing that way, with 2 concerns:
1. That way will let igniter release stale workflows if we abort a lot of workflows in the old queue. Setting the reference to the new queue at the beginning also gives the igniter ability to release workflows at the same time when the handler is enqueuing, as long as the queue object is thread-safe.
2. Having 2 queue uses more memory, when both of them are relatively big.

But revisiting it now, it looks like a better approach in terms of testing the `enqueue()`

File: falcon/queue_handler.py (hunk 20 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: Thanks, will do!

File: falcon/queue_handler.py (hunk 21 above)
Comment (dshiga) [reply falcon/queue_handler.py]: I see your point. The way you have it ordered right now might be better then. And it may not be worth passing the queue in as an argument in that case -- see how you feel when you write the unit tests to decide.

File: falcon/queue_handler.py [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+            ```
+
+        Returns:
+            bool: The return value. True if the workflow_list is sorted oldest first, False otherwise.
+        """
+        CROMWELL_DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%S.%fZ'
+        try:
+            head = datetime.strptime(str(workflow_list[0].get('submission')), CROMWELL_DATETIME_FORMAT)
+            tail = datetime.strptime(str(workflow_list[-1].get('submission')), CROMWELL_DATETIME_FORMAT)
+            return head <= tail
```
[the remark this answers is not in the document]
Comment (rexwangcc) [reply falcon/queue_handler.py]: Agree on logging. When it comes to reversing, `head == tail` I assume indicates all workflows are submitted at the same time. Are there any other reasons to not revere them in this case, except saving computation time?

[the remark this answers is not in the document]
Comment (dshiga) [reply falcon/queue_handler.py]: I was thinking it's a waste of time to reverse if the timestamps are all equal anyway. But on second thought, in practice we'll be using Falcon with the latest Cromwell where we will normally be reversing the list, so maybe it's better to keep this the way you have it, just for consistency. Seems unlikely that we'll ever hit this anyway, except maybe very occasionally when there are only two items in the list.

File: falcon/queue_handler.py [hunk 23]
```diff
@@ -0,0 +1,315 @@
...
+
+    def __init__(self, config_object):
+        self.workflow_queue = self.create_empty_queue(-1)  # use infinite for the size of the queue for now
+        self.thread = None
+        self.settings = settings.get_settings(config_object)
+        self.queue_update_interval = self.settings.get('queue_update_interval')
+        self.cromwell_url = self.settings.get('cromwell_url')
+        self.cromwell_query_dict = {
+            'status': 'On Hold',
+            'additionalQueryResultFields': 'labels',
```
Comment (samanehsan) [inline falcon/queue_handler.py]: Do you think we should request a specific number of results here? I'm not sure if we'll run into this issue, but there's a bug in Cromwell with including `additionalQueryResultFields` when the result set is over 1000 workflows: https://github.com/broadinstitute/cromwell/issues/3115

File: falcon/queue_handler.py [hunk 24]
```diff
@@ -0,0 +1,315 @@
...
+        and `bundle_version`.
+        """
+        if isinstance(other, Workflow):
+            return self.id == other.id
+        return False
+
+
+class QueueHandler(object):
+    """
+    A concrete queue handler class.
```
Comment (samanehsan) [inline falcon/queue_handler.py]: Some more context about what the queue handler is/does would be useful here, since the name already tells us that this is a concrete queue handler class! Even something brief such as "A queue handler that periodically queries Cromwell for on hold workflows after a specified time interval and adds them to a new queue."

File: falcon/queue_handler.py [hunk 25]
```diff
@@ -0,0 +1,315 @@
...
+    def execution(self):
+        logger.info(
+            'Queue | Initializing the queue handler with thread => {0} | {1}'.format(get_ident(), datetime.now()))
+
+        while True:
+            workflow_metas = self.retrieve_workflows(self.cromwell_query_dict)
+            if workflow_metas:  # This could happen when getting either non-200 codes or 0 workflow from Cromwell
+                workflows = self.prepare_workflows(workflow_metas)
+
+                self.rebuild_queue(self.create_empty_queue(-1))  # This must happen before `enqueue()` is called!
```
Comment (samanehsan) [inline falcon/queue_handler.py]: I remember when we looked at reassigning self.workflow_queue to `Queue(maxsize=max_queue_size)` and added workflows to the new queue, the igniter thought it was still empty. How does this approach get around that issue?

File: falcon/queue_handler.py (hunk 23 above)
Comment (dshiga) [reply falcon/queue_handler.py]: That is a good point. Unfortunately, since the response is now sorted newest first, we have to get all the workflows so that we can release the oldest first. We could ask for just one page of results, but that gets complicated when the sort order is the reverse of what we want. Since we are not actually using the labels for anything yet, why don't we remove `additionalQueryResultFields` for now. We will definitely have more than 1000 on hold workflows on a routine basis.

Comment (samanehsan) [reply falcon/queue_handler.py]: That's true that the change in the order of results would make requesting just one page more complicated. We'll need the bundle-uuid label for de-duplication, so we would have to add the `additionalQueryResultFields` parameter back in for that.

Comment (rexwangcc) [reply falcon/queue_handler.py]: @dshiga @samanehsan  Unfortunately, although I agree on removing the `labels` for now (That is only a placeholder which I think will be useful to de-duplicate in the future (we need both `bundle-uuid` and `bundle-version` from the `labels`to perform de-dedup)), even using `includeSubworkflows` can also lead the same error: https://github.com/broadinstitute/cromwell/issues/3873 I don't know if we will even get `On Hold` "Subworkflows" besides parent workflows, But if so(I guess no), I cannot think of a way to remove this field, since we always want to only release parent workflows.

File: falcon/queue_handler.py (hunk 25 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: This pointer movement works because I switched from `multi-processing` to `multi-threading` now :octocat: Unlike `multi-processing`, which isolates different process completely, (and the only to share objects between processes is to pickle the data to an I/O stream and simulates a Queue object), `multi-threading` lets threads share the same chunk of memory which in turn helps us do the trick here. This is like a compromise I made after our discussion last week :( 

(🥇 side-note: I will spend the same effort to learn Golang instead comparing trying different workarounds in Python, if I need to write concurrent/parallel logics next time!)

File: falcon/queue_handler.py (hunk 24 above)
Comment (rexwangcc) [reply falcon/queue_handler.py]: will do!

File: falcon/queue_handler.py (hunk 23 above)
Comment (samanehsan) [reply falcon/queue_handler.py]: That's a good point @rexwangcc! I would expect to only get `On Hold` parent workflows, but we should check what the current behavior is with the Cromwell team.

File: falcon/queue_handler.py (hunk 25 above)
Comment (samanehsan) [reply falcon/queue_handler.py]: Thanks for the explanation!

File: falcon/queue_handler.py [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+            None. It should only be spawned by the function `spawn_and_start()`.
+        settings (dict): A dictionary contains all settings for the handler.
+        queue_update_interval (int): This is the how long the handler will sleep for after each time it retrieves
+            workflows from the Cromwell.
+        cromwell_url (str): This is the Cromwell url loaded from the settings.
+        cromwell_query_dict (dict): This is the query dictionary handler uses for retrieving workflows from the
+            Cromwell. Currently this is hard-coded.
+    """
+
+    def __init__(self, config_object):
```
Comment (samanehsan) [inline falcon/queue_handler.py]: Can we rename `config_object` to `config_path`? I got confused looking at the tests when I saw a path getting passed into the QueueHandler!

Comment (rexwangcc) [reply falcon/queue_handler.py]: will do!

**Comment by rexwangcc** [comment]
@dshiga Thanks for taking a look! That sounds great, I will add a log info to the `execution_event()`!
